### ISSUE TITLE
Save feature implementation

### ISSUE THREAD

**Issue by woutergins** [woutergins/satlas#2]
Create a save and load feature for spectra, preferably with the option to save either the parameters (to reload in a Python session) or the evaluated values in a range (for plotting in a different environment).

**Comment by woutergins** [comment]
A good idea would be to also save the MLE walk data, so it can be loaded in afterwards. Adding a keyword to the correlation plot to control the burn in can speed up the determination of the burn-in time. This means we do not consider burn-in during the initial walk.

**Comment by rubendegroote** [comment]
Here is what I have in mind. Your thoughts before we go all the way with implementing it.

I would create a class that links a dataset to a (list of) spectra. This way you can keep a kind of version history of the analysis so far. You could keep track of different analysis approaches, etc. A quick suggestion for this can be found in my branch (analysis.py).

This class would be pickled to file (e.g. run263.analysis), and could be opened with a load_analysis script that unpickles and returns it, ready for further work.

**Comment by woutergins** [comment]
Not sure if the version history is strictly necessary... Might be more descriptive to change the filename for different settings of the analysis. I'm also not sure if we want to create an extra class just for saving... I would suggest adding save functionality for individual spectra, and expanding the analysis class so you can quickly generate dataframes/spreadsheets from collections of spectra.
Should the analysis class contain all the data points?

**Comment by rubendegroote** [comment]
On a separate note: pickling spectra proves quite challenging. Some numpy bugs prevent some of the steps, my lack of understanding prevents others ;).

Another route is to just save the parameter objects, and to rebuild the spectra from those objects when loading. Problem here is that the code gets very fidgety very quickly. This can and will lead to slow development and many bugs.

Perhaps the code structure would need to be altered to allow for easier saving...

**Comment by rubendegroote** [comment]
Trying a bit harder to just pickle the whole spectrum objects. It seems the problem lies in pickling the minimizer object (self.chisquare_result).

**Comment by rubendegroote** [comment]
SingleSpectrum can now be pickle dumped and loaded, but at the cost of
sacrificing two methods:
chisquare_correlation_plot
calculate_confidence_intervals

These two methods rely on the full minimizer object, and that I can't
quite pickle. All the other methods only need the parameters or the fit
report, so those are fine.

Is this worth it...?

**Comment by woutergins** [comment]
I think it is worth it. The correlation plot should be easily reproduced without access to the minimizer object. The confidence interval calculation is essentially the same thing.

**Comment by rubendegroote** [comment]
I figured that was the case. I think we can just dig in the lmfit code - I'll bet it only uses a small part of that minimizer object anyway.

I have wrapped the saving and loading code for the Analysis object into an analysisIO.py file (you'll see it's huuuuuge ;)). In my branch I also uploaded two datasets and some scripts to illustrate the usage. Just for now of course, should be removed when the branches are merged.

The walks etc. are currently not saved, and all the info is saved as a pickle dump. Next is saving those things as csv in case people want to export.

By the way, we should make an official release with all the new features once we are reasonably sure this works bug-free, and letting all the users know they should update.
